First draft of 21004 (#1318)
OpenLiberty/blogs: posts/2021-04-16-mp-context-propagration-21004.adoc

---
layout: post

> title: "MicroProfile Context Propagration 1.2 and a number of enhancements in Open Liberty 21.0.0.4"

Reviewer: "MicroProfile Context **Propagation** 1.2, **Admin Center**, and a number of enhancements in Open Liberty 21.0.0.4"

Author: Changed to ```Admin Center, MicroProfile Context Propagration 1.2, and a number of enhancements in Open Liberty 21.0.0.4```

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/jakub-pomykala
author_github: https://github.com/jakub-pomykala

> seo-title: MicroProfile Context Propagration 1.2 and a number of enhancements in Open Liberty 21.0.0.4 - OpenLiberty.io

Reviewer: "MicroProfile Context Propagation 1.2, Admin Center, and a number of enhancements in Open Liberty 21.0.0.4"

Author: Changed to ```Admin Center, MicroProfile Context Propagration 1.2, and a number of enhancements in Open Liberty 21.0.0.4```

> seo-description: In 21.0.0.4, we now offer MicroProfile Context Propagration 1.2 and a number of enhancements (expandLocation property for AutoExpand, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, Automatic Cleanup of Leaked Connections, Single Sign-On (SSO) LTPA and JWT support authentication filter and Single Sign-On (SSO) LTPA and JWT support authentication filter)

Reviewer: In 21.0.0.4, we now offer MicroProfile Context Propagation 1.2, Admin Center, and a number of enhancements.....

Author: Changed to ```In 21.0.0.4, we now offer Admin Center, MicroProfile Context Propagration 1.2, and a number of enhancements...```

> blog_description: In 21.0.0.4, we now offer MicroProfile Context Propagration 1.2 and a number of enhancements (expandLocation property for AutoExpand, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, Automatic Cleanup of Leaked Connections, Single Sign-On (SSO) LTPA and JWT support authentication filter and Single Sign-On (SSO) LTPA and JWT support authentication filter)

Reviewer: In 21.0.0.4, we now offer MicroProfile Context Propagation 1.2, Admin Center, and a number of enhancements.....

Author: Changed to ```In 21.0.0.4, we now offer Admin Center, MicroProfile Context Propagration 1.2, and a number of enhancements...```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
---

> = MicroProfile Context Propagration 1.2 and a number of enhancements in Open Liberty 21.0.0.4

Reviewer: "MicroProfile Context Propagation 1.2, Admin Center, and a number of enhancements in Open Liberty 21.0.0.4"

Author: Changed to ```Admin Center, MicroProfile Context Propagration 1.2, and a number of enhancements in Open Liberty 21.0.0.4```

Jakub Pomykala <https://github.com/jakub-pomykala>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

// tag::intro[]

> In 21.0.0.4, we now offer MicroProfile Context Propagration 1.2, which addresses the difference in how link:{url-prefix}/blog/2021/03/31/microprofile-config-2.0.html[MicroProfile Config 2.0] treats empty value configuration properties. Also in this release there is a number of enhancements which are: expandLocation property for AutoExpand, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, Automatic Cleanup of Leaked Connections, Single Sign-On (SSO) LTPA and JWT support authentication filter and Single Sign-On (SSO) LTPA and JWT support authentication filter

Reviewer: In 21.0.0.4, Admin Center 1.0 joins Open Liberty, providing a web-based UI for configuring, managing, and monitoring your server, Java Batch jobs, and OpenID Connect (OIDC). MicroProfile Context Propagation is updated to 1.2, which addresses the difference in how link:{url-prefix}/blog/2021/03/31/microprofile-config-2.0.html[MicroProfile Config 2.0] treats empty value configuration properties.  Also present in this release is a number of enhancements: expandLocation property for AutoExpand, ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, automatic cleanup of leaked connections, Single Sign-On (SSO) LTPA and JWT support authentication filter, and Single Sign-On (SSO) LTPA and JWT support authentication filter.

Author: Done

In link:{url-about}[Open Liberty] 21.0.0.4:

* <<mpContextPropagation, MicroProfile Context Propagation 1.2>>
* <<expandLocation, expandLocation property for AutoExpand>>
* <<requestTiming, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature>>
* <<JDBC, Automatic Cleanup of Leaked Connections>>
* <<SSO, Single Sign-On (SSO) LTPA and JWT support authentication filter>>

> * <<admincenter, Admin Center feature now available in Open Liberty 21.0.0.4>>

Reviewer: Move Admin Center to the first spot.  Change it to 
<<admincenter, Admin Center 1.0>>

Author: Done

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>21.0.0.4</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[21.0.0.4,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----
//end::run[]

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

//tag::features[]

[#mpContextPropagation]
== MicroProfile Context Propagation 1.2

MicroProfile Context Propagation is a standalone MicroProfile specification. MicroProfile Context Propagation enables you to create completion stages that behave deterministically with respect to thread context and leverages the autonomic tuning of the Liberty global thread pool for asynchronous dependent stages.

> The 1.2 release of MicroProfile Context Propagation aligns with the MicroProfile 4.0 platform, specifically addressing a difference in how link:{url-prefix}/blog/2021/03/31/microprofile-config-2.0.html[MicroProfile Config 2.0] treats empty value configuration properties. When using MicroProfile Config to specify an empty list of thread context types for MicroProfile Context Propagation to use as defaults, use a value of `None` rather than an empty value. An empty value in MicroProfile Config 2.0 indicates to override any lower ordinal config sources and instead use the built-in default value for the property.  For example, the combination of `mp.context.ManagedExecutor.cleared=None` and `mp.context.ManagedExecutor.propagated=Remaining` causes every context type to be propagated.

Reviewer: make "MicroProfile 4.0 platform" link to https://openliberty.io/blog/2021/03/19/microprofile40-open-liberty-21003.html

Author: Done

  <feature>servlet-4.0</feature>
  <feature>jdbc-4.2</feature>
  <feature>jndi-1.0</feature>
</featureManager>
----

Example usage within a Servlet:
[source, java]
----
private ManagedExecutor executor;

public void init(ServletConfig config) throws ServletException {
    executor = ManagedExecutor.builder()
                .propagated(ThreadContext.APPLICATION)
                .cleared(ThreadContext.ALL_REMAINING)
                .build();
}

public void destroy() {
    executor.shutdownNow();
}

public void doGet(HttpServletRequest req, HttpServletResponse resp)
    throws ServletException, IOException {
    ...
    executor.copy(unmanagedCompletionStage).thenAcceptAsync(value -> {
        // requires java:comp namespace of the application,
        DataSource ds = InitialContext.doLookup("java:comp/env/jdbc/ds");
        ...
    });
}
----

For more information please see: 
* link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2/microprofile-context-propagation-spec-1.2.html[MicroProfile Context Propagation 1.2 specification]
* link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2/apidocs/[JavaDoc]

> [#xpandLocation]

Reviewer: #expandLocation

Author: Done

[#requestTiming]
== Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature

The Request Timing feature (`requestTiming-1.0`) provides diagnostic information when the duration of any request exceeds the configured threshold. It provides a way to monitor requests with respect to time. The feature can automatically detect slow and hung requests and provide detailed diagnostic information; warning messages, thread stacks, and the creation of thread dumps.

When a hung request is detected in the Request Timing feature, a warning message is written in the messages log file along with a dump of the events that happened during the request. Following that, a set of three thread dumps will be initiated, 1 minute apart. After the completion of the three thread dumps, further set of three thread dumps are created only if new requests are detected to be hanging.

Some operations teams do not want so many thread dumps to be generated due to performance overhead on requests that are known to be long. In previous Open Liberty releases, there was no option to disable the thread dumps from being generated.

> You can now control whether the Request Timing feature collects thread dumps. By setting the **NEW** `enableThreadDumps` Request Timing server configuration attribute to false, thread dumps will not be created during hung requests. If the new server configuration attribute is set to true or not specified at all, thread dumps will still be created.

Reviewer: **new** no need for the all-caps

Author: Changed to ```**new**```

  <!-- more features -->
</featureManager>
----

Configure your data sources and connection factories as usual, which automatically leverage the new capability (To turn it off, configure `autoCloseConnections="false"` on a `<connectionManager>`).
[source, xml]
----
<dataSource id="DefaultDataSource">
  <connectionManager maxPoolSize="10"/>
  <jdbcDriver libraryRef="PostgreSQL"/>
  <properties.postgresql databaseName="TESTDB" serverName="localhost" portNumber="5432"/>
</dataSource>

<library id="PostgreSQL">
  <file name="/usr/local/postgresql/postgresql-42.2.18.jar"/>
</library>
----

Find out more at link:{url-prefix}/docs/latest/reference/config/connectionManager.html[Open Liberty Connection Manager Documentation]

[#SSO]
== Single Sign-On (SSO) LTPA and JWT support authentication filter>>

With this new enhancement users can now use the authentication filter to select which HTTP servlet request should use LTPA and JWT for the SSO authentications. To find out more about this take a look at link:{url-prefix}/blog/2020/07/02/disable-default-cookies-20007.html[Disabling default LTPA cookies for TAI/SPNEGO or default JWT cookies for JWT SSO on Open Liberty 20.0.0.7]

[#admincenter]

> == Admin Center feature now available in Open Liberty 21.0.0.4

Reviewer: Change title to "Admin Center 1.0"

Author: Done

> The adminCenter-1.0 feature enables the Liberty Administrative Center (“Admin Center”), a web-based graphical interface for deploying, monitoring and managing Liberty servers. 

Reviewer: Replace with:
The Admin Center 1.0 feature provides a web-based GUI to help configure, manage, and monitor your running server, along with support for viewing and interacting with Java Batch jobs and OpenID Connect (OIDC) client and token management.

Add the image from: /img/blog/ui_serverConfigTool2.png

In order to enable Admin Center, you'll need to add the adminCenter-1.0 feature, and an authorized user.  Since, for security reasons, Admin Center requires SSL, you'll also need to make sure you have configured the [`httpsPort`](https://openliberty.io/docs/latest/reference/config/httpEndpoint.html) and [`keyStore`](https://openliberty.io/docs/latest/reference/config/keyStore.html).

For example:
[source, xml]
----
<server description="Admin Center example">

  <!-- Enable features -->
  <featureManager>
    <feature>adminCenter-1.0</feature>
  </featureManager>

  <!-- To access this server from a remote client add a host attribute to the following element, e.g. host=“*” -->
  <httpEndpoint id=“defaultHttpEndpoint”
         host=“*”
         httpPort=“9080”
         httpsPort=“9443" />

<! -- Define a user with Administrator role -->
<quickStartSecurity userName="admin" 
userPassword="adminpwd" />

  <keyStore id=“defaultKeyStore” password=“Liberty”/>
</server>
----

Once the server starts, you'll be able to access the Admin Center login page via https://host_name:port_number/adminCenter/ so if your loopback address is mapped to localhost, you would navigate to https://localhost:9443/adminCenter/ in your browser.

Add the image from /img/blog/ui_login.png

Author: Done

> For more information about how to use this new feature, check out the link:{url-prefix}/blog/2021/04/15/admin-center-21004.html[Admin Center feature now available in Open Liberty 21.0.0.4 blog post]

Reviewer: Replace this with:
"Stay tuned for a blog specifically regarding Admin Center which will showcase the various tools and functionality it provides."

Author: Done

blog_description: In 21.0.0.4, we now offer Admin Center, MicroProfile Context Propagration 1.2, and a number of enhancements (expandLocation property for AutoExpand, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, Automatic Cleanup of Leaked Connections, Single Sign-On (SSO) LTPA and JWT support authentication filter and Single Sign-On (SSO) LTPA and JWT support authentication filter)
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Admin Center, MicroProfile Context Propagration 1.2, and a number of enhancements in Open Liberty 21.0.0.4
Jakub Pomykala <https://github.com/jakub-pomykala>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

// tag::intro[]

> In 21.0.0.4, Admin Center 1.0 joins Open Liberty, providing a web-based UI for configuring, managing, and monitoring your server, Java Batch jobs, and OpenID Connect (OIDC). MicroProfile Context Propagation is updated to 1.2, which addresses the difference in how link:{url-prefix}/blog/2021/03/31/microprofile-config-2.0.html[MicroProfile Config 2.0] treats empty value configuration properties. Also present in this release is a number of enhancements: expandLocation property for AutoExpand, ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, automatic cleanup of leaked connections, Single Sign-On (SSO) LTPA and JWT support authentication filter, and Single Sign-On (SSO) LTPA and JWT support authentication filter.

Reviewer 2: Think these may need to be edited slightly to read better:

```
expandLocation property for AutoExpand, ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, automatic cleanup of leaked connections, Single Sign-On (SSO) LTPA and JWT support authentication filter, and Single Sign-On (SSO) LTPA and JWT support authentication filter.
```

Author: Done

> In 21.0.0.4, Admin Center 1.0 joins Open Liberty, providing a web-based UI for configuring, managing, and monitoring your server, Java Batch jobs, and OpenID Connect (OIDC). MicroProfile Context Propagation is updated to 1.2, which addresses the difference in how link:{url-prefix}/blog/2021/03/31/microprofile-config-2.0.html[MicroProfile Config 2.0] treats empty value configuration properties. Also present in this release is a number of enhancements: expandLocation property for AutoExpand, ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, automatic cleanup of leaked connections, Single Sign-On (SSO) LTPA and JWT support authentication filter, and Single Sign-On (SSO) LTPA and JWT support authentication filter.

Reviewer 2: `providing a web-based UI for configuring, managing, and monitoring your server, Java Batch jobs, and OpenID Connect (OIDC).` read strangely for me the first time round, the way it's written in the `admincenter` section is much nicer imo.

Author: Changed to description from admin center section

For example:
[source, xml]
----
<server description="Admin Center example">

  <!-- Enable features -->
  <featureManager>
    <feature>adminCenter-1.0</feature>
  </featureManager>

  <!-- To access this server from a remote client add a host attribute to the following element, e.g. host=“*” -->
   <httpEndpoint id=“defaultHttpEndpoint”
         host=“*”
         httpPort=“9080”
         httpsPort=“9443" />

<! -- Define a user with Administrator role -->
<quickStartSecurity userName="admin" 
userPassword="adminpwd" />

  <keyStore id=“defaultKeyStore” password=“Liberty”/>
</server>
----

Once the server starts, you'll be able to access the Admin Center login page via https://host_name:port_number/adminCenter/ so if your loopback address is mapped to localhost, you would navigate to https://localhost:9443/adminCenter/ in your browser.

image::img/blog/ui_login.png[ui login ,width=70%,align="center"]

Stay tuned for a blog specifically regarding Admin Center which will showcase the various tools and functionality it provides.

[#mpContextPropagation]
== MicroProfile Context Propagation 1.2

> MicroProfile Context Propagation is a standalone MicroProfile specification. MicroProfile Context Propagation enables you to create completion stages that behave deterministically with respect to thread context and leverages the autonomic tuning of the Liberty global thread pool for asynchronous dependent stages.

Reviewer 2: ```suggestion
MicroProfile Context Propagation is a standalone MicroProfile specification. MicroProfile Context Propagation enables you to create completion stages that behave deterministically with respect to thread context, and leverages the autonomic tuning of the Open Liberty global thread pool for asynchronous dependent stages.
```

Author: Done

  <feature>servlet-4.0</feature>
  <feature>jdbc-4.2</feature>
  <feature>jndi-1.0</feature>
</featureManager>
----

Example usage within a Servlet:
[source, java]
----
private ManagedExecutor executor;

public void init(ServletConfig config) throws ServletException {
    executor = ManagedExecutor.builder()
                .propagated(ThreadContext.APPLICATION)
                .cleared(ThreadContext.ALL_REMAINING)
                .build();
}

public void destroy() {
    executor.shutdownNow();
}

public void doGet(HttpServletRequest req, HttpServletResponse resp)
    throws ServletException, IOException {
    ...
    executor.copy(unmanagedCompletionStage).thenAcceptAsync(value -> {
        // requires java:comp namespace of the application,
        DataSource ds = InitialContext.doLookup("java:comp/env/jdbc/ds");
        ...
    });
}
----

For more information please see: 
* link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2/microprofile-context-propagation-spec-1.2.html[MicroProfile Context Propagation 1.2 specification]
* link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2/apidocs/[JavaDoc]

> [##expandLocation]

Reviewer 2: ```suggestion
[#expandLocation]
```

Author: Done

=== expandLocation property for AutoExpand

> With this enhancement users are now able to specify a expansion location (`expandLocation`) on the `applicationManager` configuration to be utilized when the `autoExpand` attribute is set to "true". As currently implemented, when an application is autoExpanded the default location for the expanded files are hard coded to `${server.config.dir}/apps/expanded/`.

Reviewer 2: ```suggestion
With this enhancement users are now able to specify a expansion location (`expandLocation`) on the `applicationManager` configuration to be utilized when the `autoExpand` attribute is set to "true". As currently implemented, when an application is autoExpanded the default location for the expanded files is hard coded to `${server.config.dir}/apps/expanded/`.
```

Author: Think it is supposed to be "are" as it's multiple files not just one

- link:{url-prefix}/docs/latest/reference/config/requestTiming.html[Open Liberty Documentation on requestTiming Configuration]

[#JDBC]
== Automatic Cleanup of Leaked Connections

Liberty connection management is enhanced with the ability to automatically detect and close unsharable connections that are left open by the application across the end of a request.

Occasionally, application code might forget to close an unsharable connection that it obtains, which prevents the connection from being returned to the connection pool for use by other requests. Over time, these leaked connections can degrade performance and eventually exhaust the connection pool. Liberty connection management now has the ability to detect and automatically close these leaked connections to prevent this from happening.

To take advantage of this new capability, configure one of the Liberty features that leverages the `connectionManager` element. For example, JDBC:
[source, xml]
----
<featureManager>
  <feature>jdbc-4.2</feature>
  <feature>jndi-1.0</feature>
  <!-- more features -->
</featureManager>
----

> Configure your data sources and connection factories as usual, which automatically leverage the new capability (To turn it off, configure `autoCloseConnections="false"` on a `<connectionManager>`).

Reviewer 2: ```suggestion
Configure your data sources and connection factories as usual, which automatically leverage the new capability (To disable it, configure `autoCloseConnections="false"` on a `<connectionManager>`).
```

Author: Done

//end::features[]

[#bugs]
== Notable bugs fixed in this release

We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A21004+label%3A%22release+bug%22[full list of bugs fixed in 21.0.0.4].

* link:https://github.com/OpenLiberty/open-liberty/issues/16113[Shared Class Cache not generated on Windows]
+
The IBM/OpenJ9 JDK Shared Class Cache may not have been generated on Windows if `IBM_JAVA_OPTIONS/OPENJ9_JAVA_OPTIONS` was not set causing slower startup performance.  With this fix, a Shared Class Cache will be created at `<WLP_USER_DIR>/servers/.classCache` improving the startup performance.

* link:https://github.com/OpenLiberty/open-liberty/issues/16054[HSTS Header not added on responses with 404 status]
+

> In certain cases where the response is fully handled by the HTTP transport without invoking the WebContainer engine (e.g-a 404 response), the HTTP Strict-Transport-Security (`HSTS`) header was improperly omitted from the response even though the link:{url-prefix}/docs/21.0.0.3/reference/config/webContainer.html[`addStrictTransportSecurityHeader`] was properly configured.  The HTTP transport's parsing of these properties now has new tracing, the HTTP transport will inspect the response and, when configured to do so, add the `HSTS` header if it is missing and if the scheme is 'https'. The `HTTPDispatcher=all` trace level can be used to see what the resulting header value will look like.

Reviewer 2: ```suggestion
In certain cases where the response is fully handled by the HTTP transport without invoking the WebContainer engine (e.g. a 404 response), the HTTP Strict-Transport-Security (`HSTS`) header was improperly omitted from the response even though the link:{url-prefix}/docs/latest/reference/config/webContainer.html[`addStrictTransportSecurityHeader`] was properly configured.  The HTTP transport's parsing of these properties now has new tracing; the HTTP transport will inspect the response and, when configured to do so, add the `HSTS` header if it is missing. The `HTTPDispatcher=all` trace level can be used to see what the resulting header value will look like.
```

Author: Done

* link:https://github.com/OpenLiberty/open-liberty/issues/15336[Replace DNS lookup with regular expression to get the domain name in SSO Cookie Domain function]
+

> A bug in the SSO Cookie domain function caused an unnecessary DNS lookup when sending a regular HTTP Request to generate LTPA token. To improve performance, a regex will mow be used instead to obtain the domain name.

Reviewer 2: ```suggestion
A bug in the SSO Cookie domain function caused an unnecessary DNS lookup when sending a regular HTTP Request to generate an LTPA token. To improve performance, a regex will now be used instead to obtain the domain name.
```

Author: Done

* link:https://github.com/OpenLiberty/open-liberty/issues/15989[Bump netty dependencies to 4.1.62.Final]
+

> Although Open Liberty is not vulnerable, we've updated Netty (used by the link:{url-prefix}/docs/21.0.0.3/reference/feature/grpcClient-1.0.html[gRPC Client 1.0] feature) to 4.1.62.Final in order to pull in a fix for a CVE.

Reviewer 2: Did this fix not go into 21.0.0.3?

Reviewer 2: My bad this is another version update, should `/docs/21.0.0.3/reference/` -> `/docs/latest/reference/`?

Author: Doesn't seem to be in the blog

blog_description: In 21.0.0.4, we now offer Admin Center, MicroProfile Context Propagration 1.2, LDAP connection support for Kerberos authentication and a number of enhancements (expandLocation property for AutoExpand, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, Automatic Cleanup of Leaked Connections, Single Sign-On (SSO) LTPA and JWT support authentication filter and Single Sign-On (SSO) LTPA and JWT support authentication filter)
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Admin Center, MicroProfile Context Propagration 1.2, LDAP connection support for Kerberos and a number of enhancements in Open Liberty 21.0.0.4
Jakub Pomykala <https://github.com/jakub-pomykala>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

// tag::intro[]

> In 21.0.0.4, Admin Center 1.0 joins Open Liberty, providing a web-based GUI to help configure, manage, and monitor your running server, along with support for viewing and interacting with Java Batch jobs and OpenID Connect (OIDC) client and token management. MicroProfile Context Propagation is updated to version 1.2, which addresses the difference in how link:{url-prefix}/blog/2021/03/31/microprofile-config-2.0.html[MicroProfile Config 2.0] treats empty value configuration properties. LDAP connection now supports Kerberos authentication. Also present in this release is a number of enhancements: expandLocation property for AutoExpand, the ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, automatic cleanup of leaked connections, Single Sign-On (SSO) LTPA and JWT support authentication filter, and Single Sign-On (SSO) LTPA and JWT support authentication filter

Reviewer 2: ```suggestion
In 21.0.0.4, Admin Center 1.0 joins Open Liberty, providing a web-based GUI to help configure, manage, and monitor your running server, along with support for viewing and interacting with Java Batch jobs and OpenID Connect (OIDC) client and token management. MicroProfile Context Propagation is updated to version 1.2, which addresses the difference in how link:{url-prefix}/blog/2021/03/31/microprofile-config-2.0.html[MicroProfile Config 2.0] treats empty value configuration properties. LDAP connection now supports Kerberos authentication. Also present in this release are a number of enhancements: expandLocation property for AutoExpand, the ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature, automatic cleanup of leaked connections, Single Sign-On (SSO) LTPA and JWT support authentication filter, and Single Sign-On (SSO) LTPA and JWT support authentication filter
```

Author: Done
